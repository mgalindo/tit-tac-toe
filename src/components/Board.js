import React from "react";
import PropTypes from "prop-types";
import { View, StyleSheet, Text } from "react-native";
import Box from "./Box";

const Board = props => {
  const renderRow = rowNum => {
    const columns = [1, 2, 3];
    const { curPlayer, values, onPlayerMove } = props;
    return columns.map(colNum => {
      const index = colNum + 3 * rowNum;
      return (
        <Box
          key={index}
          index={index}
          value={values[index - 1]}
          curPlayer={curPlayer}
          onPlayerMove={onPlayerMove}
        />
      );
    });
  };

  const rows = [0, 1, 2];
  const { curPlayer } = props;
  return (
    <View style={styles.boardContainer}>
      {rows.map(rowNum => {
        return (
          <View key={rowNum.toString()} style={styles.row}>
            {renderRow(rowNum)}
          </View>
        );
      })}
      <Text style={styles.playerText} testID="player-text">
        Next Player: {curPlayer}
      </Text>
    </View>
  );
};

Board.propTypes = {
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
  curPlayer: PropTypes.string.isRequired,
  onPlayerMove: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
  boardContainer: {
    flex: 0.8,
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center"
  },
  playerText: {
    marginTop: 25,
    fontSize: 30,
    color: "#4A8305"
  },
  row: {
    flexDirection: "row"
  }
});

export default Board;
