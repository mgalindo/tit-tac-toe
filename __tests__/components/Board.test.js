import React from "react";
import { render, fireEvent } from "react-native-testing-library";

import Board from "../../src/components/Board";

describe("Board component", () => {
  let props;
  let onMockPlayerMove;

  beforeEach(() => {
    onMockPlayerMove = jest.fn();
    jest.clearAllMocks();
    props = {
      values: new Array(9).fill(" "),
      curPlayer: "X",
      onPlayerMove: onMockPlayerMove
    };
  });

  test("should render empty board", () => {
    expect.assertions(2);
    const { toJSON, getAllByTestId } = render(<Board {...props} />);
    const boxes = getAllByTestId("box-button");

    expect(toJSON()).toMatchSnapshot();
    expect(boxes.length).toBe(9);
  });

  test("should be able to click all boxes on empty board", () => {
    expect.assertions(1);
    const { getAllByTestId } = render(<Board {...props} />);
    const boxes = getAllByTestId("box-button");

    fireEvent(boxes[0], "press");
    fireEvent(boxes[1], "press");
    fireEvent(boxes[2], "press");
    fireEvent(boxes[3], "press");
    fireEvent(boxes[4], "press");
    fireEvent(boxes[5], "press");
    fireEvent(boxes[6], "press");
    fireEvent(boxes[7], "press");
    fireEvent(boxes[8], "press");

    expect(onMockPlayerMove).toHaveBeenCalledTimes(9);
  });

  test("should display the next player", () => {
    expect.assertions(1);
    const { getByTestId } = render(<Board {...props} />);
    const nextPlayerText = getByTestId("player-text");
    expect(nextPlayerText.props.children).toEqual([
      "Next Player: ",
      props.curPlayer
    ]);
  });

  test("should render full board", () => {
    expect.assertions(2);
    props.values = ["X", "O", "X", "O", "X", "O", "X", "O", "X"];
    const { toJSON, getAllByTestId } = render(<Board {...props} />);
    const boxes = getAllByTestId("box-button");

    expect(toJSON()).toMatchSnapshot();
    expect(boxes.length).toBe(9);
  });
});
