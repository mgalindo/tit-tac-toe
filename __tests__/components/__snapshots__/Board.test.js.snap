// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component should render empty board 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 0.8,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
           
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#4A8305",
        "fontSize": 30,
        "marginTop": 25,
      }
    }
    testID="player-text"
  >
    Next Player: 
    X
  </Text>
</View>
`;

exports[`Board component should render full board 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 0.8,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "grey",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          X
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          O
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "grey",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          X
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          O
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "grey",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          X
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          O
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "grey",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          X
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          O
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#4A8305",
            "height": 150,
            "justifyContent": "center",
            "marginBottom": 5,
            "marginLeft": 6,
            "width": 150,
          },
          Object {
            "backgroundColor": "grey",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID="box-button"
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
            }
          }
          testID="box-text"
        >
          X
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#4A8305",
        "fontSize": 30,
        "marginTop": 25,
      }
    }
    testID="player-text"
  >
    Next Player: 
    X
  </Text>
</View>
`;
